src/components/tl/mlx5/mcast/tl_mlx5_mcast_team.c

```diff
@@ -128,3 +298,295 @@ ucc_status_t ucc_tl_mlx5_mcast_coll_setup_comm_resources(ucc_tl_mlx5_mcast_coll_
+                net_addr.sin6_flowinfo = comm->comm_id;
+                status = ucc_tl_mlx5_mcast_join_mcast_post(comm->ctx, &net_addr, 1);
+                if (status < 0) {
+                    tl_error(comm->lib, "rank 0 is unable to join mcast group error %d", status);
+                    tl_team->mcast_state = TL_MLX5_TEAM_STATE_MCAST_GRP_JOIN_FAILED;
+                    return UCC_INPROGRESS;
+                }
+
+                tl_team->mcast_state = TL_MLX5_TEAM_STATE_MCAST_GRP_JOIN_POST;
+
+                return UCC_INPROGRESS;
+            }
+
+            case TL_MLX5_TEAM_STATE_MCAST_GRP_JOIN_POST:
+            {
+                /* rank 0 has already called rdma_join_multicast() 
+                 * it is time to wait for the rdma event to confirm the join */
+                status = ucc_tl_mlx5_mcast_join_mcast_test(comm->ctx, &comm->event, 1);
+                if (UCC_OK != status) {
+                    tl_team->mcast_state = TL_MLX5_TEAM_STATE_MCAST_GRP_JOIN_FAILED;
+                    if (comm->event) {
+                        rdma_ack_cm_event(comm->event);
+                    }
+                    return UCC_INPROGRESS;
+                }
+
+                ucc_assert(comm->event != NULL);
+
+                /* at this point, rank 0 has joined mcast group */
+                tl_team->mcast_state = TL_MLX5_TEAM_STATE_MCAST_GRP_JOIN_READY;
+
+                return UCC_INPROGRESS;
+            }
+
+            case TL_MLX5_TEAM_STATE_MCAST_GRP_JOIN_READY:
+            {
+                /* now rank 0 bcast the lid/gid to other processes */
+                data = ucc_malloc(sizeof(ucc_status_t) + mgid_s + mlid_s);
+                if (!data) {
+                    tl_error(comm->lib, "unable to allocate memory for group setup info");
+                    return UCC_ERR_NO_MEMORY;
+                }
+
+                comm->group_setup_info = data;
+                memset(data, 0, sizeof(ucc_status_t) + mgid_s + mlid_s);
+                memcpy((PTR_OFFSET(data, sizeof(ucc_status_t))),
+                        &comm->event->param.ud.ah_attr.grh.dgid, mgid_s);
+                memcpy((PTR_OFFSET(data, sizeof(ucc_status_t) + mgid_s)),
+                        &comm->event->param.ud.ah_attr.dlid, mlid_s);
+                memcpy(data, &status, sizeof(ucc_status_t));
+
+                status = comm->bcast_post(comm->p2p_ctx, data, sizeof(ucc_status_t) + mgid_s +
+                                          mlid_s, 0, &comm->group_setup_info_req);
+                if (UCC_OK != status) {
+                    tl_error(comm->lib, "unable to post bcast for group setup info");
+                    ucc_free(comm->group_setup_info);
+                    if (comm->event) {
+                        rdma_ack_cm_event(comm->event);
+                    }
+                    return status;
+                }
+
+                comm->mcast_lid  = *((uint16_t*)(PTR_OFFSET(data, sizeof(ucc_status_t) +
```

Comment:
It is dynamic allocation as data is the buffer which is used in Bcast. Since rank 0 (root) is  the one running this code line, then it is safe to use data's buffer content for setting the lid/gid. However, non-roots must wait until bcast is completed to do the same thing. FYI root is sending out the info about gid/lid/status to nonroots here.